torchvision/models/regnet.py

```diff
@@ -0,0 +1,785 @@
+        # and activation is not counted with respect to depth
+        self.depth = self.f.depth
+
+    def forward(self, x: Tensor) -> Tensor:
+        if self.proj_block:
+            x = self.bn(self.proj(x)) + self.f(x)
+        else:
+            x = x + self.f(x)
+        return self.activation(x)
+
+
+class ResBottleneckLinearBlock(nn.Module):
+    """Residual linear bottleneck block: x + F(x), F = bottleneck transform."""
+
+    def __init__(
+        self,
+        width_in: int,
+        width_out: int,
+        stride: int,
+        bn_epsilon: float,
+        bn_momentum: float,
+        activation: nn.Module,
+        group_width: int = 1,
+        bottleneck_multiplier: float = 4.0,
+        se_ratio: Optional[float] = None,
+    ) -> None:
+        super().__init__()
+        self.has_skip = (width_in == width_out) and (stride == 1)
+        self.f = BottleneckTransform(
+            width_in,
+            width_out,
+            stride,
+            bn_epsilon,
+            bn_momentum,
+            activation,
+            group_width,
+            bottleneck_multiplier,
+            se_ratio,
+        )
+
+        self.depth = self.f.depth
+
+    def forward(self, x: Tensor) -> Tensor:
+        return x + self.f(x) if self.has_skip else self.f(x)
+
+
+class AnyStage(nn.Sequential):
+    """AnyNet stage (sequence of blocks w/ the same output shape)."""
+
+    def __init__(
+        self,
+        width_in: int,
+        width_out: int,
+        stride: int,
+        depth: int,
+        block_constructor: Callable[..., nn.Module],
+        bn_epsilon: float,
+        bn_momentum: float,
+        activation: nn.Module,
+        group_width: int,
+        bottleneck_multiplier: float,
+        se_ratio: Optional[float] = None,
+        stage_index: int = 0,
+    ) -> None:
+        super().__init__()
+        self.stage_depth = 0
+
+        for i in range(depth):
+            block = block_constructor(
+                width_in if i == 0 else width_out,
+                width_out,
+                stride if i == 0 else 1,
+                bn_epsilon,
+                bn_momentum,
+                activation,
+                group_width,
+                bottleneck_multiplier,
+                se_ratio,
+            )
+
+            self.stage_depth += block.depth
+            self.add_module(f"block{stage_index}-{i}", block)
+
+
+class RegNetParams:
+    def __init__(
+        self,
+        depth: int,
+        w_0: int,
+        w_a: float,
+        w_m: float,
+        group_width: int,
+        bottleneck_multiplier: float = 1.0,
+        stem_type: Callable[..., nn.Module] = SimpleStemIN,
+        stem_width: int = 32,
+        block_type: Callable[..., nn.Module] = ResBottleneckBlock,
+        activation: Callable[..., nn.Module] = nn.ReLU,
+        use_se: bool = True,
+        se_ratio: float = 0.25,
+        bn_epsilon: float = 1e-05,
+        bn_momentum: float = 0.1,
+        num_classes: int = 1000,
```

Comment:
The majority of the constructors in the models of vision receive many of these params (such as `num_classes`) in the constructor. See [here](https://github.com/pytorch/vision/blob/main/torchvision/models/googlenet.py#L68).

Reply:
I'm also considering whether we should have this kind of params container. If so, what should be in the params container?
Would it make more sense if we only keep the params that are used in `get_expanded_params` in the container and move the rest(e.g. num_classes, stem_type, block_type, activation) to model constructor?

Reply:
Very good points.

The latter approach worked very nicely form me in the past (aka passing the rest on constructor). I also started constructing the parameter objects in the building methods and passing them on the constructor (instead of adding them on a method of the class). That allowed me to parameterize and change the architecture of the network at will with minimum modifications on the Model class itself.

Now to be clear, the above approach is not taken universally in the codebase but rather followed on latest models such as MobileNetV3 and EfficientNets. So perhaps this is something we should discuss and decide here for going forwards. No strong opinions on my side. @fmassa any thoughts around this?